src/components/CanvasSaveButton/CanvasSaveButton.js

```diff
@@ -61,53 +98,138 @@ class CanvasSaveButton extends PureComponent {
     );
   }
 
-  getCanvasImage(opts, asMSBlob) {
-    const { map } = this.props;
-    let image;
+  createCanvasImage(opts, asMSBlob) {
+    return new Promise(resolve => {
+      const { map, extent, extraData } = this.props;
+      let image;
 
-    map.once('postcompose', evt => {
-      const { canvas } = evt.context;
-      const { height, width } = canvas;
+      map.once('postcompose', evt => {
+        const { canvas } = evt.context;
 
-      const destinationCanvas = document.createElement('canvas');
-      destinationCanvas.width = width;
-      destinationCanvas.height = height;
+        let clip = {
+          x: 0,
+          y: 0,
+          w: canvas.width,
+          h: canvas.height,
+        };
 
-      const destContext = destinationCanvas.getContext('2d');
-      destContext.fillStyle = 'white';
-      destContext.fillRect(0, 0, width, height);
-      destContext.drawImage(canvas, 0, 0, width, height, 0, 0, width, height);
+        if (extent) {
+          const pixelTopLeft = map.getPixelFromCoordinate(getTopLeft(extent));
+          const pixelBottomRight = map.getPixelFromCoordinate(
+            getBottomRight(extent),
+          );
 
-      if (asMSBlob) {
-        image = destinationCanvas.msToBlob();
-      } else {
-        image = destinationCanvas.toDataURL(opts.format);
-      }
+          clip = {
+            x: pixelTopLeft[0],
+            y: pixelTopLeft[1],
+            w: pixelBottomRight[0] - pixelTopLeft[0],
+            h: pixelBottomRight[1] - pixelTopLeft[1],
+          };
+        }
+
+        const destCanvas = document.createElement('canvas');
+        destCanvas.width = clip.w;
+        destCanvas.height = clip.h;
+        const destContext = destCanvas.getContext('2d');
+
+        // Draw map
+        destContext.fillStyle = 'white';
+        destContext.fillRect(0, 0, destCanvas.width, clip.h);
+        destContext.drawImage(
+          canvas,
+          clip.x,
+          clip.y,
+          clip.w,
+          clip.h,
+          0,
+          0,
+          clip.w,
+          clip.h,
+        );
+
+        const padding = 5;
+        const arrowSize = 80;
+
+        // Copyright
+        if (extraData && extraData.copyright) {
+          destContext.font = '12px Arial';
+          destContext.fillStyle = 'black';
+          destContext.fillText(extraData.copyright, padding, clip.h - padding);
+        }
+
+        // North arrow
+        if (extraData && extraData.northArrow) {
+          const img = new Image();
+          img.src = extraData.northArrow.circled
```

Comment:
What if we pass a child to NorthArrow, to use it as custom north arrow, such as in this example:
https://github.com/geops/react-spatial/blob/master/src/components/NorthArrow/NorthArrow.test.js#L29
Do you think we should support this scenario here as well?

Reply:
he's right,`northArrow` to display  should come from the properties of the object as well.

Reply:
> Changes I propose :
>     * [...]
>     * Add properties src, width, height to northArrow property to be able to specify a custom northArrow.

Done with src, width, and height